packages/compose/src/hooks/use-ref-effect/index.ts

```diff
@@ -17,14 +23,17 @@ import { useCallback, useRef } from '@wordpress/element';
  * to be removed. It *is* necessary if you add dependencies because the ref
  * callback will be called multiple times for the same node.
  *
- * @param {Function} callback     Callback with ref as argument.
- * @param {Array}    dependencies Dependencies of the callback.
+ * @param callback Callback with ref as argument.
+ * @param dependencies Dependencies of the callback.
  *
- * @return {Function} Ref callback.
+ * @return Ref callback.
  */
-export default function useRefEffect( callback, dependencies ) {
-	const cleanup = useRef();
-	return useCallback( ( node ) => {
+export default function useRefEffect< TElement = Node >(
+	callback: ( node: TElement ) => ( () => void ) | undefined,
```

Comment:
@sarayourfriend: can we use `EffectCallback` now? Like:

```js
callback: ( node: TElement ) => EffectCallback,
```

Reply:
I don't believe so because that would imply that the callback's return is also able to return a desctructor function which will go one level too deep I think?

Here's the definition for `EffectCallback`:
```ts
type EffectCallback = () => (void | (() => void | undefined));
```

If we used `EffectCallback` then the cleanup function, which is the current optional return of the `callback` parameter, would itself also be allowed to return a function, but we completely ignore the return type of the cleanup function.

If `EffectCallback` was a generic type that allowed us to add an input parameter type, then we could use it like this:

```ts
callback: EffectCallback<TElement>
```

But effects take no input for their callback so that will never be the case.

I could be confused here though, let me know if you're seeing something I'm not.

Reply:
Yeah, you're totally right, I was the one confused. I misread some type definitions in a different module and thought we were missing an opportunity to use a predefined type. Thanks for humouring me. :)